src/thunderbots/software/util/visualizer_messenger/visualizer_messenger.cpp

```diff
@@ -42,169 +87,156 @@ namespace Util
+    void VisualizerMessenger::clearShapes()
     {
-        ShapeMsg new_shape;
-        new_shape.type = "arc";
-        new_shape.data.clear();
-        new_shape.data.push_back(cx);
-        new_shape.data.push_back(cy);
-        new_shape.data.push_back(radius);
-        new_shape.data.push_back(theta_start);
-        new_shape.data.push_back(theta_end);
-
-        applyDrawStyleToMsg(new_shape, draw_style);
-        applyDrawTransformToMsg(new_shape, draw_transform);
-        addShapeToLayer(layer, new_shape);
+        for (auto& layer : this->layer_shapes_map)
+        {
+            layer.second.clear();
+        }
     }
 
-    void VisualizerMessenger::drawLine(const std::string& layer, double x1, double y1,
-                                       double x2, double y2, DrawStyle draw_style,
-                                       DrawTransform draw_transform)
+    void VisualizerMessenger::drawEllipse(uint8_t layer, uint16_t cx, uint16_t cy,
+                                          int16_t r1, int16_t r2, int16_t rotation,
+                                          ShapeStyle style)
     {
-        ShapeMsg new_shape;
-        new_shape.type = "line";
-        new_shape.data.clear();
-        new_shape.data.push_back(x1);
-        new_shape.data.push_back(y1);
-        new_shape.data.push_back(x2);
-        new_shape.data.push_back(y2);
-
-        applyDrawStyleToMsg(new_shape, draw_style);
-        applyDrawTransformToMsg(new_shape, draw_transform);
-        addShapeToLayer(layer, new_shape);
-    }
+        const uint8_t texture = style.texture;
+        const uint32_t tint   = style.tint;
 
-    void VisualizerMessenger::buildLayers()
-    {
-        // TODO: #268 list these existing layer names elsewhere where it's more obvious
-        std::vector<std::string> layer_names = std::vector<std::string>(
-            {"field", "ball", "ball_vel", "friendly_robot", "friendly_robot_vel",
-             "enemy_robot", "enemy_robot_vel", "nav", "rule", "passing", "test", "misc"});
+        Shape new_shape;
+        new_shape.texture = texture;
 
-        for (std::string layer_name : layer_names)
-        {
-            LayerMsg new_layer_msg;
-            new_layer_msg.layer_name = layer_name;
-            this->layers_name_to_msg_map.insert(
-                std::pair<std::string, LayerMsg>(layer_name, new_layer_msg));
-        }
+        // Since x and y of the shape definition is top left, we need to shift to respect
+        // the center of the ellipse
+        new_shape.x = cx - r1;
+        new_shape.y = cy - r2;
+
+        new_shape.width  = r1 * 2;
+        new_shape.height = r2 * 2;
+
+        new_shape.rotation = rotation;
+        new_shape.tint     = tint;
+
+        this->addShapeToLayer(layer, new_shape);
     }
 
-    void VisualizerMessenger::applyDrawStyleToMsg(ShapeMsg& shape_msg, DrawStyle& style)
+    void VisualizerMessenger::drawRect(uint8_t layer, int16_t x, int16_t y, int16_t w,
```

Comment:
If we can only draw predefined textures, not necessarily rectangles, should we instead have a general methods to draw sprites instead of specific shapes? Or should we still have those methods that point to specific texture numbers in the spritesheet?

Reply:
Just jumping in here, I think that it makes sense to draw specific shapes like this, because it reduces the API for drawing so that there are less requirements we need to fulfill if we ever need to go to another frontend (ie. this reduces coupling). This also ensures we're doing nothing wrong at compile time, which is super nice.

Reply:
Just jumping in here, I think that it makes sense to draw specific shapes like this, because it reduces the API for drawing so that there are less requirements we need to fulfill if we ever need to go to another frontend (ie. this reduces coupling). This also ensures we're doing nothing wrong at compile time, which is super nice.